Add image to QuestionStep and FormStep, support imageAltText on any step containing image

eschramm opened the pull request:
Closes https://github.com/CareEvolution/MyDataHelps-iOS/issues/1158. Adds `image` to ORK[1]QuestionStep and ORK[1]FormStep and `imageAltText ` to all steps that support `image`.

The latter implies the use of Accessibility (a11y) to “read” a representation of the image as supplied in `imageAltText`. This required some of the significant lift in this PR as a11y was not working as expected on Question- and FormSteps.

Support for a11y here is defined as being able to tap on the image with Settings > Accessibility > Spoken Content turned on as well as being able to cycle the views on a screen and see them spoken in order (can be done with Apple’s Accessibility Inspector using Xcode and a simulator, or with Spoken Content on, two-finger swipe down on screen).

Due to low usage (only used by one study to my knowledge) and spending several hours struggling with it, this PR does NOT provide full a11y for ORKQuestionsStep (RK2) when a UITableView is used (e.g., TextChoices). The question `text` in RK2 is actually placed in a Section Header which is not a standard place for a11y when looking at UITableview hierarchy. It might be achieved by passing more data around or after a deep dive on subclassing aspects of UITableView.

### RK1
|QuestionStep|FormStep|
|---|---|
|![Simulator Screen Shot - iPhone 14 - 2023-03-06 at 14 48 10](https://user-images.githubusercontent.com/1008462/223227757-4ab9d4e7-673c-4907-8c44-0634caccece2.png)|![Simulator Screen Shot - iPhone 14 - 2023-03-06 at 14 48 33](https://user-images.githubusercontent.com/1008462/223227794-3c72fa84-fff5-4480-9cee-7a45444e578e.png)|

### RK2
|QuestionStep|FormStep|
|---|---|
|![Simulator Screen Shot - iPhone 14 - 2023-03-06 at 14 49 28](https://user-images.githubusercontent.com/1008462/223227901-402a101e-f654-474e-9439-2f256a3709d7.png)|![Simulator Screen Shot - iPhone 14 - 2023-03-06 at 14 48 56](https://user-images.githubusercontent.com/1008462/223227934-b1785de9-eb64-4bd9-850f-c4abe8d6e4c7.png)|

## Testing

Until [this server PR](https://github.com/CareEvolution/MyDataHelpsDesignerUI/pull/20) is merged/live, this can only be tested by modifying survey JSON via the Edit Raw Template. The surveys (1 each for RK version 1 and RK version 2) below have a QuestionStep and an InstructionStep, both with images.

To test `altImageText`, you can turn on Spoken Content either by triple-tapping the power button or via Settings > Accessibility > Spoken Content. Then you can tap on the image to hear the alternate text.

Expectations 
- Images appear after title and text, but before answer section in both the QuestionStep and Instruction Step
- When in Spoken Content mode on an iPhone, each image will have a description: QuestionStep - MDH logo - "MyDataHelps logo", FormStep - Framingham Heart Study logo - "eFHS logo"

<details>
  <summary>RK1 Question/Form Step Survey example</summary>

```
{
  "type": "RKStudioOrderedTask",
  "name": "Survey",
  "identifier": "Survey",
  "steps": [
    {
      "type": "QuestionStep",
      "text": "This is text. Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Pellentesque habitant morbi tristique senectus.",
      "title": "Question Step",
      "identifier": "QuestionStep",
      "image": "https://play-lh.googleusercontent.com/NvRkVn1T5D1rjUryUdsn74p3TzfzesXER8SijXnwKezppm5m9OvwafeSx0jr4QF4uw",
      "imageAltText": "MyDataHelps logo",
      "answerFormat": {
        "type": "TextAnswerFormat",
        "multipleLines": false,
        "keyboardType": "Text"
      },
      "optional": false,
      "placeholder": "This is the placeholder",
      "useCardView": true
    },
    {
      "type": "FormStep",
      "identifier": "FormStep",
      "image": "https://rkstudio-customer-assets.s3.amazonaws.com/OFFSPRING/three-heart-logo.gif",
      "imageAltText": "eFHS logo",
      "formItems": [
        {
          "identifier": "FormItem1",
          "type": "FormItem",
          "answerFormat": {
            "type": "TextAnswerFormat",
            "multipleLines": false,
            "keyboardType": "Text"
          },
          "optional": false,
          "text": "Text Question",
          "placeholder": "placeholder"
        },
        {
          "identifier": "FormItem2",
          "type": "FormItem",
          "answerFormat": {
            "type": "NumericAnswerFormat",
            "style": "Integer"
          },
          "optional": false,
          "text": "Integer"
        }
      ],
      "optional": false,
      "title": "Form Step",
      "text": "This is text. Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Pellentesque habitant morbi tristique senectus."
    }
  ]
}
```

</details>
<details>
  <summary>RK2 Question/Form Step Survey example</summary>

```
{
  "type": "RKStudioOrderedTask",
  "name": "Survey",
  "identifier": "Survey",
  "rkVersion": 2,
  "steps": [
    {
      "type": "QuestionStep",
      "text": "This is text. Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Pellentesque habitant morbi tristique senectus.",
      "title": "Question Step",
      "identifier": "QuestionStep",
      "image": "https://play-lh.googleusercontent.com/NvRkVn1T5D1rjUryUdsn74p3TzfzesXER8SijXnwKezppm5m9OvwafeSx0jr4QF4uw",
      "imageAltText": "MyDataHelps logo",
      "answerFormat": {
        "type": "TextAnswerFormat",
        "multipleLines": false,
        "keyboardType": "Text"
      },
      "optional": false,
      "placeholder": "This is the placeholder",
      "useCardView": true
    },
    {
      "type": "FormStep",
      "identifier": "FormStep",
      "image": "https://rkstudio-customer-assets.s3.amazonaws.com/OFFSPRING/three-heart-logo.gif",
      "imageAltText": "eFHS logo",
      "formItems": [
        {
          "identifier": "FormItem1",
          "type": "FormItem",
          "answerFormat": {
            "type": "TextAnswerFormat",
            "multipleLines": false,
            "keyboardType": "Text"
          },
          "optional": false,
          "text": "Text Question",
          "placeholder": "placeholder"
        },
        {
          "identifier": "FormItem2",
          "type": "FormItem",
          "answerFormat": {
            "type": "NumericAnswerFormat",
            "style": "Integer"
          },
          "optional": false,
          "text": "Integer"
        }
      ],
      "optional": false,
      "title": "Form Step",
      "text": "This is text. Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Pellentesque habitant morbi tristique senectus."
    }
  ]
}
```

</details>

eschramm committed:
add support for image/altImageText for Question/FormSteps

eschramm committed:
fix order of accessibility items so image is before customView

mmertsock commented on ORK1Kit/ORK1Kit/Common/ORK1TableContainerView.h:
```diff
@@ -54,6 +54,8 @@ NS_ASSUME_NONNULL_BEGIN
 @property (nonatomic, strong, readonly) UITableView *tableView;
 @property (nonatomic, strong, readonly) ORK1StepHeaderView *stepHeaderView;
 @property (nonatomic, strong, readonly) ORK1NavigationContainerView *continueSkipContainerView;
+@property (nonatomic, strong) UIImage *image;
```
Minor: image and imageAltText properties should have `nullable` annotation

eschramm replied:
Good catch!

mmertsock commented on ORK1Kit/ORK1Kit/Common/ORK1InstructionStepView.m:
```diff
@@ -137,7 +137,11 @@ - (void)setInstructionStep:(ORK1InstructionStep *)instructionStep {
         
         _instructionImageView.isAccessibilityElement = YES;
-        _instructionImageView.accessibilityLabel = [NSString stringWithFormat:ORK1LocalizedString(@"AX_IMAGE_ILLUSTRATION", nil), _instructionStep.title];
+        if (_instructionStep.imageAltText) {
```
Do you know whether our survey JSON will definitely have a nil value for imageAltText when the survey designer hasn't entered anything? Or would it be an empty string (in which case, this if-statement would not work as intended)

eschramm replied:
I do not, but I agree we could be defensive here and cover for that case. This should be updated now here and in other similar places.

mmertsock commented on ResearchKit/Common/ORKTableContainerView.h:
```diff
@@ -53,6 +53,8 @@ NS_ASSUME_NONNULL_BEGIN
 
 @property (nonatomic, strong, readonly) UITableView *tableView;
 @property (nonatomic, strong, readonly) ORKStepHeaderView *stepHeaderView;
+@property (nonatomic, strong) UIImage *image;
```
Same here: should have `nullable` annotation (to override the `NS_ASSUME_NONNULL_BEGIN` above)

mmertsock reviewed (approved):
LGTM 🚢 
Just some minor comments.

eschramm committed:
PR fixes

- defensively handling empty string vs nil
- making new image properties nullable

mmertsock reviewed (approved):
Latest changes LGTM 🚢
